Fix jsonfile cache plugin adds trailing space at end of lines with commas

cuongnv23 opened the pull request:
##### ISSUE TYPE

 - Bugfix Pull Request


##### COMPONENT NAME
lib/ansible/parsing/utils/jsonify.py

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 61e6e7868c) last updated 2017/01/21 09:04:05 (GMT +700)
  config file = 
  configured module search path = Default w/o overrides

```

##### SUMMARY
Use `jsonfile` cache plugin with configuration:
```
gathering = smart
fact_caching = jsonfile
fact_caching_connection = /tmp/test
```

The cache file contains trailing space with lines end with comma (`,`), like this:

![json_dump](http://i.imgur.com/Ib6q99N.png)

(the `¬` character represents the end of line)

This PR adds `separators` to handle that trailing space. Also simplifies the unit test.

After fixed:
![fixed](http://i.imgur.com/6eFLB0E.png)

mwedgwood-rmn committed:
Add account ID grouping and attribute to ease multi-account management

cuongnv23 committed:
Fix json.dumps adds trailing space end of line

cuongnv23 committed:
Fix json.dumps adds trailing space end of line

Commit:
updates ensure_connect decorator function

only calls _connect() if _connected is False

cuongnv23 committed:
Simplify jsonify unit test

MarkMielke committed:
Reduce the system calls performed by the "file" module
when using "state: link", and particularly when using
"force: yes".

Symbolic link resolution can be expensive. In our case,
the symbolic links are legacy links to automounts, and
the "file" task was causing all of the legacy links to
be traversed and mounted on every host every time the
task executed, even when the links were correct and there
was nothing to do.

This change avoids the system calls that perform the
symbolic link resolution by taking advantage of the short
circuit behaviur of the boolean "and" operator. The code
behaviour is unchanged except that it no longer performs
unnecessary system calls.

As it turns out, this change is not sufficient to fully
solve the symbolic link resolution problem, as the "file"
module still performs a stat() at the end of execution to
provide the caller with information about the file.
However, this change is very simple, it will eliminate
unnecessary system calls in a number of use cases, and it
gets the "file" module closer to the desired end result.

privateip committed:
fixes minor issue in get_block_config() (#20544)

passed an invalid argument to dumps() to return the config block

machacekondra committed:
Add oVirt modules notes to CHANGELOG (#20401)

bcoca commented:
what does this break?

bcoca commented on test/units/parsing/utils/test_jsonify.py:
```diff
@@ -27,9 +27,7 @@ def test_jsonify_simple(self):
-        res = jsonify(dict(a=1, b=2, c=3), format=True)
-        cleaned = "".join([x.strip() for x in res.splitlines()])
-        self.assertEqual(cleaned, '{"a": 1,"b": 2,"c": 3}')
+        self.assertEqual(jsonify(dict(a=1, b=2, c=3), format=True), '{\n    "a":1,\n    "b":2,\n    "c":3\n}')
```
both are valid tests, i don't see why you need to eliminate the existing one

cuongnv23 replied:
yes, both are valid with my PR, as my commit message says, this just simplifies as I don't see the advantage of making the code more complicated.

bcoca committed:
updated timezone docs to avoid repainting bikeshed

fixes #19272
fixes #20570

bcoca committed:
actually fix tickets in previous commit

both modules had similar issues and i conflated them

bcoca committed:
added formatting

dbarrosop committed:
First release of the callback selective.py (#17309)

tntC4stl3 committed:
Support option none in docker_container when setting log_driver (#19040)

Per official docker document, it support setting `--log-driver=none` to
disable any logging for the container. So let's add it to this module.

Fixes #5337

Qalthos committed:
vyos_command 2.3 (#18994)

* Peter's new module_utils/vyos
* Update vyos_command
* Restore `set terminal length $ANSIBLE_VYOS_TERMINAL_LENGTH` in vyos
* vyos_command tests
* Remove provider mentions from vyos_command
* Extend get_config to take (and cache) various `show configuration` commands

mattclay committed:
Fix git remote url change detection. (#20587)

* Fix git remote url change detection.

When comparing paths, unfrackgitpath must be called on both sides
of the comparison. Otherwise comparisons involving symlinks will
return incorrect results.

* Re-enable git test on OS X.

* Return None from unfrackgitpath when path is None.

cuongnv23 commented:
@bcoca current code adds unnecessary spaces, I'm sure that we don't need that

resmo committed:
cloudstack: cs_instance: fix ssh key always marked as changed (#20541)

The same ssh key may be registered with different names. Before we only looked at the names and marked as changed when names differ. However internally cloudstack seems to ignore the change, if the fingerprints are identical. 

As a result we also have to compare the fingerprints and only mark the keys as changed if the fingerprints differ.

dagwieers committed:
win_psexec: execute cmds on remote systems as any user (#20141)

* win_psexec: execute cmds on remote systems as any user

This module uses the versatile psexec tool to run any command remotely
as any user (incl. domain users).

* Add missing documentation

Now that this module is deemed acceptable for inclusion,
the documentation is an essential part.

* win_psexec: Small cosmetic changes

* win_psexec: add more options (priority, elevated, ...)

* Fixes after more testing

* Renamed 'cmd' to 'psexec_command' + more

- Also replaced PSObject() with a hash table
- Made $chdir of type "path"
- Renamed $args to $extra_args

* Various improvements

- Switched to using booleans for most parameters
- Added type 'bool' to boolean parameters
- Added 'interactive' parameter
- Added 'wait' parameter
- Added an interactive example

* Added -type "bool" support to Get-AnsibleParam

* Fix deadlock

* When using `wait:no` return code is PID of process

naslanidis committed:
new AWS module for ec2 VPC vgw facts (#19021)

* new AWS module for ec2 VPC vgw facts

* fixed documentation as suggested by reviewers

* changed iteritems to items for py 3

* improvements to bring in line with #19787

* corrected documentation

* updated doco and exception path

ryansb committed:
Add ec2_vpc_vgw_facts module to changelog

bcoca commented:
i still fail to see why this is a problem.

Also why one test is better than the other? I'm fine with adding a new one, but not with removing the old one as they are different tests.

dagwieers committed:
win_robocopy: Small cleanup (#20515)

- Harmonizing the checkmode usage in windows modules
- Simplify the $result object
- Add parameter types

nerzhul committed:
Add logstash_plugin to manange logstash plugins (#20592)

resmo committed:
cloudstack: fix user data unexpected marked as changed (#20540)

self.instance may be overwritten before the user data comparsion which result in a false positive changed state. 

With this commit, we make sure we have the userdata when we need to.

rcarrillocruz committed:
Fix typo on test_ios_config_before_after_no_change test (#20565)

anryko committed:
Fixed bug in ecs_taskdefinition for new task with forced revision 1. (#20561)

dminca committed:
update GCE instance image with Debian 8 Jessie (#20558)

* update instance image with Debian 8 Jessie

* debian-7 is marked as "DEPRECATED" in Google Compute Engine Images
  * as a result, by default use `debian-8` Jessie

Resolves:
Related:
Signed-off-by: Daniel Andrei Minca <danielandrei.minca@gameloft.com>

* update docs regarding latest stable default image

* after speaking with @gundalow, we decided it's better for the users to
  know that the default image will follow latest stable debian by GCE

Resolves: #20558
Related:
Signed-off-by: Daniel Andrei Minca <danielandrei.minca@gameloft.com>

mmaslowskicc committed:
lambda: Remove the note about supporting S3 uploads only (#20562)

The module supports also ZIP uploads which are documented.

Refer to the correct boto3 function in the return value description.

aperigault committed:
vmware_guest: Amazing speed up module (#19937)

* vmware_guest: Amazing speed up module
* vmware_guest: Change variables name to do PEP compliant

jonathan-fugue committed:
Updating credstash link to the new link (#20598)

We changed the name of the github repo awhile back and I noticed it was out of date in the docs.

mattclay committed:
Add startup checks to ansible-test.

bcoca committed:
docs fixes

fixed version number
updated C year

Commit:
Increase timeout for http server in uri test.

nitzmahone committed:
winrm managed kinit (#20416)

cuongnv23 commented:
I think what we don't need, we should remove. I'm ok to keep the old test, let me remove last commit and update this PR. Thanks

cuongnv23 committed:
Merge branch 'fix_jsondump' of github.com:cuongnv23/ansible into fix_jsondump

Conflicts:
	lib/ansible/parsing/utils/jsonify.py

cuongnv23 commented:
my git messed up, close this one, let me send another pr.
